core/src/main/java/org/apache/carbondata/core/scan/executor/util/RestructureUtil.java

```diff
@@ -119,12 +122,15 @@
         for (CarbonDimension tableDimension : tableComplexDimension) {
           if (isColumnMatches(isTransactionalTable, queryDimension.getDimension(),
               tableDimension)) {
-            ProjectionDimension currentBlockDimension = null;
+            ProjectionDimension currentBlockDimension;
             // If projection dimension is child of struct field and contains Parent Ordinal
             if (null != queryDimension.getDimension().getComplexParentDimension()) {
               currentBlockDimension = new ProjectionDimension(queryDimension.getDimension());
             } else {
               currentBlockDimension = new ProjectionDimension(tableDimension);
+              newComplexChildrenDatatypes = new HashMap<>();
+              fillNewDatatypesForComplexChildren(queryDimension.getDimension());
+              compareDatatypes(currentBlockDimension.getDimension());
```

Comment:
Instead of creating a map, made changes to directly call another function to get dimension and set datatype. As we need to get dimension by traversing to the child columns, a recursive method is necessary.